Added excercise mode with questions from learning materials

ND-SR7 opened the pull request:
Found a fork of this repo using questions from the LPI learning material and thought it might be good to include them in the original. They were written in German and since there are a lot of them (around 200), I used a bulk translator and went over the questions once to verify them and check for bad translations. Further work might be needed to improve the quality of questions. Some questions could maybe be dropped. I would recommend someone else take a look at the questions before merging it on the main branch.

Also added a simple switch between exam and exercise mode, based on URL. Exam mode uses already existing questions, while the exercise mode uses new batch of questions from the learning materials. Feel free to adjust mode switching in any way you see fit.

On a different note, I would recommend adding a license to the repo, since it is quite popular and to prevent any offshoot from taking all the credit. If you don't care about licensing, just adding a simple MIT license would be more than enough.

Noam-Alum commented:
It will take me sometime to check this out since I don't have access to my computer in the next few days, thank you very much!

Can you send me a link to that fork?

This project exists *just because* so licensing isn't top of mind, but I'll add one as soon as I can since it seems right to do so.

The idea to add a switch is really good! I used an old version of this website a while ago to create other exams, and so I added some more data in the JSON file like the exam name, exam description, etc... to create them dynamically, we might want to expand in this direction instead of a switch (I haven't seen it yet, I'm sure it's awesome 🙃).
I recall someone reaching to me asking for another practice exam to be made for another certification, this could be a good community project where everyone add to those JSON files. (Then the multiple exams idea will make more sense)

ND-SR7 commented:
It's this [fork](https://github.com/JernicVR/lpi_010_160_theoretical_exam).

> ... to create them dynamically, we might want to expand in this direction instead of a switch

We can work on it in any way you see fit, I just submitted something simple and functional to go along with the PR, in case someone wants to use those additional questions immediately.

> I recall someone reaching to me asking for another practice exam to be made for another certification, this could be a good community project where everyone add to those JSON files.

Sounds like a great idea, I think it would only require few tweaks to the JSON and we could have a dedicated template for creating exams.

Noam-Alum commented:
Let me know what you think, then we can merge this to *main*, but I think we'll have to update the README file as well.

Can you swap it with this on your branch:

[![](https://img.shields.io/github/sponsors/Noam-Alum?label=Sponsor&message=%E2%9D%A4&logo=GitHub&color=%23fe8e86)](https://github.com/sponsors/Noam-Alum)

# LPI 010-160 practice exam

The Linux Essentials certificate validates a demonstrated understanding of the following: FOSS, the various communities, and licenses. Knowledge of open source applications in the workplace as they relate to closed source equivalents.

<p align="center">
  <img src="https://www.lpi.org/wp-content/uploads/2023/04/Essentials-Linux_250_0.png" alt="alt text">
</p>

This bash script/[webpage](https://noam-alum.github.io/lpi_010_160_exam/) lets you practice for LPI 010-160 (Linux Essentials) with real questions from the exam.

<hr>

# Take the exam online

You can take the **EXACT** exam online [here](https://noam-alum.github.io/lpi_010_160_exam/).

![Online test](https://noam-alum.github.io/lpi_010_160_exam/hero.webp)

With extensive results:

![results-page](https://noam-alum.github.io/lpi_010_160_exam/results-page.png)

Or even an *extended* exam for folks who want to learn more 😉, by adding `extended` to the exam url, or by clicking [here](https://noam-alum.github.io/lpi_010_160_exam/extended).

**This includes:**

- Random question order.
- Random answer choices order.
- Full fleged feedback at the end of the exam.

> Feel free to submit a [feature request](https://github.com/Noam-Alum/lpi_010_160_exam/issues/new?template=feature_request.md).


>[!NOTE]
> ### If this helped you, consider [sponsering](https://github.com/sponsors/Noam-Alum) this repository 😁.


# Installation

## Linux:
### wget
```bash
wget -O lpi_010_160_exam.zip https://codeload.github.com/Noam-Alum/lpi_010_160_exam/zip/refs/heads/main
unzip lpi_010_160_exam.zip
cd lpi_010_160_exam-main/
chmod +x lpi.sh
```
### git clone
```bash
git clone --single-branch --branch main --depth 1 https://github.com/Noam-Alum/lpi_010_160_exam.git
cd lpi_010_160_exam/
chmod +x lpi.sh
```

## Windows:

> If you want to use this bash script on Windows and you have *[WSL](https://blogs.windows.com/windowsdeveloper/2016/03/30/run-bash-on-ubuntu-on-windows/)* you can follow the instructions for **Linux**.

I tested this with [Git bash](https://git-scm.com/downloads), and it works well, first you would have to install **Git bash** on your computer, then you'd have to install [jq](https://github.com/jqlang/jq/releases) for windows and put the exe file in a folder so it can to be included in the $PATH variable, e.g. `[GIT BASH FOLDER]\mingw64\bin` (Usually `C:\Program Files\Git`), finally you can run the script just make sure its has the execute bit on.

### git clone
```bash
git clone --single-branch --branch main --depth 1 https://github.com/Noam-Alum/lpi_010_160_exam.git
cd lpi_010_160_exam/
chmod +x lpi.sh

# ! Make sure you followed the instructions above !
```

<br>
<hr>

# Usage

You can execute the script by simply running:
```bash
./lpi.sh
```
> You must have **jq** installed!

<br>

Then you should get the first question of the bunch, e.g. :
```
noam ◈ noam ⊛ lpi_010_160_exam ⊛ ❯❯ ./lpi.sh 


    ██╗     ██████╗ ██╗    ███████╗██╗  ██╗ █████╗ ███╗   ███╗
    ██║     ██╔══██╗██║    ██╔════╝╚██╗██╔╝██╔══██╗████╗ ████║
    ██║     ██████╔╝██║    █████╗   ╚███╔╝ ███████║██╔████╔██║
    ██║     ██╔═══╝ ██║    ██╔══╝   ██╔██╗ ██╔══██║██║╚██╔╝██║
    ███████╗██║     ██║    ███████╗██╔╝ ██╗██║  ██║██║ ╚═╝ ██║
    ╚══════╝╚═╝     ╚═╝    ╚══════╝╚═╝  ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝


-ˋˏ✄┈┈┈┈ Setting environment ...

 ➣ Validating url.
 ➣ Fetching LPI questions from  https://alum.sh/files/lpi/lpi_questions.json .

-ˋˏ✄┈┈┈┈ Done! ʘ‿ʘ


LPI practice exam:
━━━━━━ʕ•㉨•ʔ━━━━━━━

 ➣ Which of the following tar options handle compression? (Choose two.)

   1) -z
   2) -g
   3) -z2
   4) -bz
   5) -j

   ➣ Answer number 1 (1 - 5):
```

---

## Contributors

Huge thanks for contributing to this project — I really appreciate it!

#### [ND-SR7](https://github.com/ND-SR7)

- [Added the extended exam option](https://github.com/Noam-Alum/lpi_010_160_exam/pull/17)  
- [Implemented the fill-in-the-blank questions feature](https://github.com/Noam-Alum/lpi_010_160_exam/pull/16)  
- [Fixed a UI issue](https://github.com/Noam-Alum/lpi_010_160_exam/pull/15)  
- And for being the very first person to submit a PR 🙃

ND-SR7 commented:
I've updated the README, hope I didn't miss anything. Also updated the URL switch to use `/extended`.

Have you thought about licensing? Maybe we can push that with this PR or leave it for later

Noam-Alum commented:
Oh yeah totally forgot about that, what license should we use? MIT?

I actually remove the note from the README file, you can turn the comments here to plane text and just copy paste 😉 sorry for making you do all that work.

ND-SR7 commented:
MIT if you want people to do whatever but must attribute you. Alternatively a type of copyleft license, if you want all subsequent forks to remain free and open source. AGPL might be good for this, since this app can be deployed online. But I think MIT would suffice.

I've adjusted the README, I can't find the option for viewing comments in plain text

Noam-Alum commented:
> MIT if you want people to do whatever but must attribute you. Alternatively a type of copyleft license, if you want all subsequent forks to remain free and open source. AGPL might be good for this, since this app can be deployed online. But I think MIT would suffice.
> 
> I've adjusted the README, I can't find the option for viewing comments in plain text

I just hopped on the computer, seems like you'r right, I thought you could view raw comments 🤔.  (Funny enough it seems like [this](https://api.github.com/repos/Noam-Alum/lpi_010_160_exam/issues/17/comments) is the only way to get it)

I think MIT would work fine for this project, once you add it we can merge this to main.

Noam-Alum reviewed (commented):
Looks good.

ND-SR7 commented on docs/src/App.jsx:
```diff
@@ -21,7 +21,13 @@ const App = () => {
-        fetch("https://raw.githubusercontent.com/Noam-Alum/lpi_010_160_exam/refs/heads/main/lpi/lpi_questions.json")
+        let fetchUrl = "https://raw.githubusercontent.com/Noam-Alum/lpi_010_160_exam/refs/heads/main/lpi/";
+        
+        window.location.href.includes("/lpi_010_160_exam/practice") ?
```
I have changed it to `extended`, it appears you are viewing an outdated version

Noam-Alum commented:
Alright, looks good! Thank you very much.
